Extractor/Deletes.cs

```diff
@@ -0,0 +1,108 @@
+﻿using Cognite.Extractor.StateStorage;
+using Cognite.OpcUa.NodeSources;
+using Microsoft.Extensions.Logging;
+using Opc.Ua;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Cognite.OpcUa
+{
+    internal class NodeExistsState : BaseExtractionState
+    {
+        public NodeExistsState(string id) : base(id)
+        {
+        }
+    }
+
+    public class DeletedNodes
+    {
+        public IEnumerable<string> Objects { get; }
+        public IEnumerable<string> Variables { get; }
+        public IEnumerable<string> References { get; }
+        public DeletedNodes(IEnumerable<string> objects, IEnumerable<string> variables, IEnumerable<string> references)
+        {
+            Objects = objects;
+            Variables = variables;
+            References = references;
+        }
+
+        public DeletedNodes()
+        {
+            Objects = Enumerable.Empty<string>();
+            Variables = Enumerable.Empty<string>();
+            References = Enumerable.Empty<string>();
+        }
+    }
+
+    public class DeletesManager
+    {
+        private readonly IExtractionStateStore stateStore;
+        private readonly IUAClientAccess client;
+        private readonly FullConfig config;
+        private readonly ILogger logger;
+        public DeletesManager(IExtractionStateStore stateStore, IUAClientAccess client, ILogger<DeletesManager> logger, FullConfig config)
+        {
+            this.stateStore = stateStore;
+            this.client = client;
+            this.config = config;
+            this.logger = logger;
+        }
+
+        private async Task<IEnumerable<string>> GetDeletedItems(string? tableName, Dictionary<string, NodeExistsState> states, CancellationToken token)
+        {
+            // If there are no states we don't check for deletes. This means we can't ever delete the _last_ of a type, but that is likely to be
+            // an issue or a bug either way.
+            if (tableName == null || !states.Any()) return Enumerable.Empty<string>();
+
+            Dictionary<string, BaseStorableState> oldStates;
+            try
+            {
+                oldStates = (await stateStore.GetAllExtractionStates<BaseStorableState>(tableName, token)).ToDictionary(s => s.Id);
+            }
+            catch (Exception ex)
+            {
+                logger.LogWarning(ex, "Failed to get states from state store, assuming it is empty");
+                oldStates = new Dictionary<string, BaseStorableState>();
+            }
+            var deleted = oldStates.Where(s => !states.ContainsKey(s.Key)).Select(kvp => kvp.Value).ToList();
```

Comment:
Rather than creating a dangling variable, why don't we extract this into its own method (maybe `GetDeletedItems`). Makes it easier to read too.

Reply:
That seems a little unpleasant? We have to pass both `states` and `oldStates` to that method, and it will be a one-liner. I don't really see how that is clearer?

Reply:
I gave it a shot, but it takes up twice as many lines of code, does not reduce duplication, and IMO does not make the code more readable.

Reply:
The method can return both the deleted and the new items so it represents a sort of separation of states.

Reply:
I'm not sure I understand what you are asking for?